src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl_SourceGenerators.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+
+internal abstract partial class VisualStudioWorkspaceImpl
+{
+    private readonly AsyncBatchingWorkQueue _updateSourceGeneratorsQueue;
+    private bool _isSubscribedToSourceGeneratorImpactingEvents;
+
+    public void SubscribeToSourceGeneratorImpactingEvents()
+    {
+        _foregroundObject.AssertIsForeground();
+        if (_isSubscribedToSourceGeneratorImpactingEvents)
+            return;
+
+        // UIContextImpl requires IVsMonitorSelection service:
+        if (ServiceProvider.GlobalProvider.GetService(typeof(IVsMonitorSelection)) == null)
+            return;
+
+        _isSubscribedToSourceGeneratorImpactingEvents = true;
+
+        // This pattern ensures that we are called whenever the build starts/completes even if it is already in progress.
+        KnownUIContexts.SolutionBuildingContext.WhenActivated(() =>
+        {
+            KnownUIContexts.SolutionBuildingContext.UIContextChanged += (_, e) =>
+            {
+                if (!e.Activated)
+                {
+                    // After a build occurs, transition the solution to a new source generator version.  This will
+                    // ensure that any cached SG documents will be re-generated.
+                    this.EnqueueUpdateSourceGeneratorVersion();
+                }
+            };
+        });
+
+        KnownUIContexts.SolutionExistsAndFullyLoadedContext.WhenActivated(() =>
+        {
+            KnownUIContexts.SolutionExistsAndFullyLoadedContext.UIContextChanged += (_, e) =>
+            {
+                if (e.Activated)
+                {
+                    // After the solution fully loads, transition the solution to a new source generator version.  This
+                    // will ensure that we'll now produce correct SG docs with fully knowledge of all the user's state.
+                    this.EnqueueUpdateSourceGeneratorVersion();
+                }
+            };
+        });
```

Comment:
this ensures that once the PS has computed everything and told us about everything we need, that we ensure generators run.